### ISSUE TITLE
spark on ray cluster failed

### ISSUE THREAD

**Issue by Second222None** [oap-project/raydp#190]
Ray cluster works and Spark is only deployed on the master node. When the task is scheduled to the master node, everything is right. When the task is scheduled to the worker node, an error occurred. Details are as follows. Can anyone give me some instructions? Thanks!
```
(base) [root@master ray]# ray status
======== Cluster status: 2021-11-01 15:02:45.515356 ========
Node status
------------------------------------------------------------
 1 node(s) with resources: {'memory': 130950863872.0, 'node:90.91.1.163': 1.0, 'CPU': 80.0, 'object_store_memory': 60407513088.0}
 1 node(s) with resources: {'object_store_memory': 60532752384.0, 'CPU': 80.0, 'node:90.91.1.172': 1.0, 'memory': 141243088896.0}

Resources
------------------------------------------------------------
Usage:
 0.0/160.0 CPU
 0.00/253.500 GiB memory
 0.00/112.634 GiB object_store_memory
```

Run the following code:
```
import ray
import raydp

ray.init(address='auto')

@ray.remote
class PySparkDriver:
  def __init__(self):
    assert ray.is_initialized()
    self.spark = raydp.init_spark(
        app_name='RayDP example',
        num_executors=10,
        executor_cores=1,
        executor_memory="2GB")

  def foo(self):
    return self.spark.range(1000000).repartition(1000).count()

  def stop(self):
    raydp.stop_spark()

driver = PySparkDriver.remote()
print(ray.get(driver.foo.remote()))
ray.get(driver.stop.remote())
```
The following error occured:
```
(base) [root@master ray]# python spark_on_cluster.py
2021-11-01 14:53:50,985 INFO worker.py:640 -- Connecting to existing Ray cluster at address: 90.91.1.163:6379
(pid=None, ip=90.91.1.172) Setting default log level to "WARN".
(pid=None, ip=90.91.1.172) To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Traceback (most recent call last):
  File "spark_on_cluster.py", line 23, in <module>
    print(ray.get(driver.foo.remote()))
  File "/root/anaconda3/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/root/anaconda3/lib/python3.8/site-packages/ray/worker.py", line 1483, in get
    raise value
ray.exceptions.RayActorError: The actor died because of an error raised in its creation task, ray::PySparkDriver.__init__() (pid=39356, ip=90.91.1.172)
  File "python/ray/_raylet.pyx", line 505, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 449, in ray._raylet.execute_task.function_executor
  File "/root/anaconda3/lib/python3.8/site-packages/ray/_private/function_manager.py", line 556, in actor_method_executor
    return method(__ray_actor, *args, **kwargs)
  File "spark_on_cluster.py", line 10, in __init__
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 122, in init_spark
    return _global_spark_context.get_or_create_session()
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 68, in get_or_create_session
    spark_cluster = self._get_or_create_spark_cluster()
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 62, in _get_or_create_spark_cluster
    self._spark_cluster = SparkCluster(self._configs)
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster.py", line 33, in __init__
    self._set_up_master(None, None)
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster.py", line 39, in _set_up_master
    self._app_master_bridge.start_up()
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster_master.py", line 59, in start_up
    self._create_app_master(extra_classpath)
  File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster_master.py", line 169, in _create_app_master
    self._app_master_java_bridge.startUpAppMaster(extra_classpath)
  File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1304, in __call__
    return_value = get_return_value(
  File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/protocol.py", line 334, in get_return_value
    raise Py4JError(
py4j.protocol.Py4JError: An error occurred while calling o0.startUpAppMaster
(pid=None, ip=90.91.1.172) ERROR:root:Exception while sending command.
(pid=None, ip=90.91.1.172) Traceback (most recent call last):
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1207, in send_command
(pid=None, ip=90.91.1.172)     raise Py4JNetworkError("Answer from Java side is empty")
(pid=None, ip=90.91.1.172) py4j.protocol.Py4JNetworkError: Answer from Java side is empty
(pid=None, ip=90.91.1.172)
(pid=None, ip=90.91.1.172) During handling of the above exception, another exception occurred:
(pid=None, ip=90.91.1.172)
(pid=None, ip=90.91.1.172) Traceback (most recent call last):
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1033, in send_command
(pid=None, ip=90.91.1.172)     response = connection.send_command(command)
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1211, in send_command
(pid=None, ip=90.91.1.172)     raise Py4JNetworkError(
(pid=None, ip=90.91.1.172) py4j.protocol.Py4JNetworkError: Error while receiving
(pid=None, ip=90.91.1.172) 2021-11-01 14:54:13,014      ERROR worker.py:382 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::PySparkDriver.__init__() (pid=39356, ip=90.91.1.172)
(pid=None, ip=90.91.1.172)   File "python/ray/_raylet.pyx", line 505, in ray._raylet.execute_task
(pid=None, ip=90.91.1.172)   File "python/ray/_raylet.pyx", line 449, in ray._raylet.execute_task.function_executor
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/ray/_private/function_manager.py", line 556, in actor_method_executor
(pid=None, ip=90.91.1.172)     return method(__ray_actor, *args, **kwargs)
(pid=None, ip=90.91.1.172)   File "spark_on_cluster.py", line 10, in __init__
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 122, in init_spark
(pid=None, ip=90.91.1.172)     return _global_spark_context.get_or_create_session()
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 68, in get_or_create_session
(pid=None, ip=90.91.1.172)     spark_cluster = self._get_or_create_spark_cluster()
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/context.py", line 62, in _get_or_create_spark_cluster
(pid=None, ip=90.91.1.172)     self._spark_cluster = SparkCluster(self._configs)
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster.py", line 33, in __init__
(pid=None, ip=90.91.1.172)     self._set_up_master(None, None)
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster.py", line 39, in _set_up_master
(pid=None, ip=90.91.1.172)     self._app_master_bridge.start_up()
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster_master.py", line 59, in start_up
(pid=None, ip=90.91.1.172)     self._create_app_master(extra_classpath)
(pid=None, ip=90.91.1.172)   File "/root/anaconda3/lib/python3.8/site-packages/raydp/spark/ray_cluster_master.py", line 169, in _create_app_master
(pid=None, ip=90.91.1.172)     self._app_master_java_bridge.startUpAppMaster(extra_classpath)
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1304, in __call__
(pid=None, ip=90.91.1.172)     return_value = get_return_value(
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/protocol.py", line 334, in get_return_value
(pid=None, ip=90.91.1.172)     raise Py4JError(
(pid=None, ip=90.91.1.172) py4j.protocol.Py4JError: An error occurred while calling o0.startUpAppMaster
(pid=None, ip=90.91.1.172) ERROR:py4j.java_gateway:An error occurred while trying to connect to the Java server (127.0.0.1:25333)
(pid=None, ip=90.91.1.172) Traceback (most recent call last):
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 977, in _get_connection
(pid=None, ip=90.91.1.172)     connection = self.deque.pop()
(pid=None, ip=90.91.1.172) IndexError: pop from an empty deque
(pid=None, ip=90.91.1.172)
(pid=None, ip=90.91.1.172) During handling of the above exception, another exception occurred:
(pid=None, ip=90.91.1.172)
(pid=None, ip=90.91.1.172) Traceback (most recent call last):
(pid=None, ip=90.91.1.172)   File "/root/package/spark-3.0.3-bin-hadoop2.7/python/lib/py4j-0.10.9-src.zip/py4j/java_gateway.py", line 1115, in start
(pid=None, ip=90.91.1.172)     self.socket.connect((self.address, self.port))
(pid=None, ip=90.91.1.172) ConnectionRefusedError: [Errno 111] Connection refused
```

**Comment by carsonwang** [comment]
You need pySpark and Java installed on every node. If you are using conda, please make sure the same environment is available on every node.

**Comment by Second222None** [comment]
> You need pySpark and Java installed on every node. If you are using conda, please make sure the same environment is available on every node.

I checked my worker node environment. 
```
(base) [root@worker1 ~]# pip list |grep spark
pyspark                            3.0.3
(base) [root@worker1 ~]# java -version
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
```
Just now, I have deployed Spark on the worker node and I can create a Spark session successfully using the following code.
```
import ray
import raydp

ray.init()
spark = raydp.init_spark(
  app_name = "example",
  num_executors = 10,
  executor_cores = 64,
  executor_memory = "256GB"
)
```
The information on dashboard is also correct.
![image](https://user-images.githubusercontent.com/24448002/139639368-016cbe23-443c-46ec-8edd-f63e0cbf2eec.png)

**Comment by kira-lin** [comment]
Is the problem solved? What do you mean by deploy spark?

**Comment by Second222None** [comment]
I'm not sure if I need a Spark cluster. But a Spark cluster does not help. The same error.

**Comment by kira-lin** [comment]
You do not need a traditional Spark cluster. When you run raydp.init_spark, executors will be started as actors in the ray cluster. If the ray cluster contains multuple nodes, then executors are probably distributed among these nodes, This is controlled by ray. When does the program fails, and when does it work? Can you please check if it works when you directly run raydp.init_spark in command line on each node?

**Comment by Second222None** [comment]
It seems to be nothing wrong. 
On master node:
```
(base) [root@master ray]# python
Python 3.8.8 (default, Apr 13 2021, 19:58:26)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
>>> import raydp
>>> ray.init()
2021-11-01 17:01:43,078 INFO services.py:1267 -- View the Ray dashboard at http://127.0.0.1:8266
{'node_ip_address': '90.91.1.163', 'raylet_ip_address': '90.91.1.163', 'redis_address': '90.91.1.163:30432', 'object_store_address': '/tmp/ray/session_2021-11-01_17-01-41_526624_61466/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2021-11-01_17-01-41_526624_61466/sockets/raylet', 'webui_url': '127.0.0.1:8266', 'session_dir': '/tmp/ray/session_2021-11-01_17-01-41_526624_61466', 'metrics_export_port': 40103, 'node_id': '7cdcfa75e7115dca3e3d5d0f73bc0414cad4537777cd7c0b817cb015'}
>>> spark = raydp.init_spark(
...   app_name = "example",
...   num_executors = 10,
...   executor_cores = 64,
...   executor_memory = "256GB"
... )
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
2021-11-01 17:02:02 WARN  NativeCodeLoader:60 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
21/11/01 17:02:04 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
>>>
```

On worker node:
```
(base) [root@worker1 ~]# python
Python 3.8.8 (default, Apr 13 2021, 19:58:26)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
>>> import raydp
>>> ray.init()
2021-11-01 17:04:55,937 INFO services.py:1267 -- View the Ray dashboard at http://127.0.0.1:8265
{'node_ip_address': '90.91.1.172', 'raylet_ip_address': '90.91.1.172', 'redis_address': '90.91.1.17       2:6379', 'object_store_address': '/tmp/ray/session_2021-11-01_17-04-54_516793_55684/sockets/plasma_       store', 'raylet_socket_name': '/tmp/ray/session_2021-11-01_17-04-54_516793_55684/sockets/raylet', '       webui_url': '127.0.0.1:8265', 'session_dir': '/tmp/ray/session_2021-11-01_17-04-54_516793_55684', '       metrics_export_port': 65376, 'node_id': '3a61c7aa6f46f139dacb7ac9db4bf50208758e42a663f632af2653a3'}
>>> spark = raydp.init_spark(
...   app_name = "example",
...   num_executors = 10,
...   executor_cores = 64,
...   executor_memory = "256GB"
... )
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
2021-11-01 17:05:21 WARN  NativeCodeLoader:60 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
21/11/01 17:05:23 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
>>>
```

**Comment by kira-lin** [comment]
Okay. You can try to specify which node to run the actor by following:
```python
@ray.remote(resources={"node:90.91.1.163": 0.1})
class PySparkDriver:
```
And see when does it fail. I just tried it in my cluster, and it works on all nodes. BTW, if it still fails, maybe you can first try our 0.4.0 version, which is going to release soon.

**Comment by Second222None** [comment]
When `raydp.init_spark` is scheduled to the remote node, it failed.

I can see the first print result("before init_spart") and then the error information.

```
@ray.remote(resources={"node:90.91.1.172": 0.1})
class PySparkDriver:
  def __init__(self):
    assert ray.is_initialized()
    print("before init_spart")
    self.spark = raydp.init_spark(
        app_name='RayDP example',
        num_executors=10,
        executor_cores=10,
        executor_memory="2GB")
    print("init_spark successfully")
```

**Comment by kira-lin** [comment]
Did you install pyspark via pip? Or did you have only spark binary installation on both nodes?

**Comment by Second222None** [comment]
Yes, I installed **ray-1.3.0**, **raydp-0.3.0**, and **pyspark-3.0.3** via pip on both nodes. And **spark-3.0.3-bin-hadoop2.7** and **hadoop-3.1.0** is installed on master node.

**Comment by kira-lin** [comment]
I see. In the error message, it shows the spark binary location, which is not on the worker. On master, in the environment you call raydp.init_spark, if you import pyspark, what is pyspark.__file__? I guess it points to the binary installation. Can you try to make this point to the same location on the two nodes? That should solve the problem.

**Comment by Second222None** [comment]
You are right. By default, **Spark** uses the pyspark in its own path on master node. On the worker node, there is  no **Spark** installed.
Now I remove **Spark** on the master node.  On both nodes, **pyspark.__file__** points to **/root/anaconda3/lib/python3.8/site-packages/pyspark/__init__.py** . 
On both nodes, the following code runs successfully:
```
import ray
import raydp

ray.init()
spark = raydp.init_spark(
  app_name = "example",
  num_executors = 10,
  executor_cores = 64,
  executor_memory = "256GB"
)
```
But the code I mentioned above reports the following error when I start it on worker node:
```
(base) [root@worker1 ray]# python spark_on_cluster.py
/root/anaconda3/lib/python3.8/site-packages/pyspark/__init__.py
2021-11-03 11:27:47,343 INFO worker.py:640 -- Connecting to existing Ray cluster at address: 90.91.1.163:6379
Aborted (core dumped)
```
I have no idea about it.

**Comment by kira-lin** [comment]
Can you be more specific on what causes core dump? Which line of code? By the way, are you using raydp 0.3.0? We have released raydp 0.4.0 yesterday, and please use ray 1.7.0 with raydp 0.4.0, not ray 1.8.0 which is released this morning.

**Comment by Second222None** [comment]
Problem solved with **ray-1.7.0** and **raydp-0.4.0**. 
Thank you!
```
(base) [root@master ray]# pip list | grep ray
bitarray                           2.1.0
ray                                1.7.0
raydp                              0.4.0
```
